File: java/src/main/java/it/giannibombelli/wsc2026/payment/infrastructure/providers/GiftCard.java
package it.giannibombelli.wsc2026.payment.infrastructure.providers;

import it.giannibombelli.wsc2026.common.domain.primitive.Description;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.common.domain.primitive.Timestamp;
import it.giannibombelli.wsc2026.payment.domain.ports.PaymentProvider;
import it.giannibombelli.wsc2026.payment.domain.ports.PaymentProviderResult;

import java.util.UUID;

public final class GiftCard implements PaymentProvider {
    @Override
    public PaymentProviderResult process(UUID paymentId, UUID providerReference, Money amount) {
        if (providerReference == null) {
            return new PaymentProviderResult.Failure(new Description("gift card reference required"));
        }

        return new PaymentProviderResult.Success(UUID.randomUUID(), Timestamp.now());
    }

    @Override
    public PaymentProviderResult refund(UUID paymentId, UUID providerReference, Money amount) {
        if (providerReference == null) {
            return new PaymentProviderResult.Failure(new Description("gift card reference required"));
        }

        return new PaymentProviderResult.Success(UUID.randomUUID(), Timestamp.now());
    }
}

File: typescript/src/payment/infrastructure/providers/giftCard.ts
// Adapter per il provider GiftCard (stub che richiede un riferimento).

import { Description } from '@/common/domain/primitive/description.js';
import { Timestamp } from '@/common/domain/primitive/timestamp.js';
import { Uuid } from '@/common/domain/primitive/uuid.js';
import type { Money } from '@/common/domain/primitive/money.js';
import type { PaymentProvider } from '../../domain/ports/paymentProvider.js';
import {
  PaymentProviderResult,
  paymentProviderFailure,
  paymentProviderSuccess,
} from '../../domain/ports/paymentProviderResult.js';

export class GiftCard implements PaymentProvider {
  process(_paymentId: Uuid, providerReference: Uuid, _amount: Money): PaymentProviderResult {
    if (providerReference === null) {
      return paymentProviderFailure(new Description('gift card reference required'));
    }

    return paymentProviderSuccess(Uuid.generate(), Timestamp.now());
  }

  refund(_paymentId: Uuid, providerReference: Uuid, _amount: Money): PaymentProviderResult {
    if (providerReference === null) {
      return paymentProviderFailure(new Description('gift card reference required'));
    }

    return paymentProviderSuccess(Uuid.generate(), Timestamp.now());
  }
}
